File: ios/FinanceApp/Reports/ReportsViewModel.swift
import Foundation
import SwiftUI
import Combine

class ReportsViewModel: ObservableObject {
    @Published var exportState: ExportState = .idle
    @Published var reportState: ReportState = .idle
    @Published var loanTrackerState: LoanTrackerState = .idle
    
    private let repository: FinanceRepository
    private let excelExporter: ExcelExporter
    
    init(repository: FinanceRepository, excelExporter: ExcelExporter) {
        self.repository = repository
        self.excelExporter = excelExporter
    }
    
    func generateReport(startDate: Date, endDate: Date) {
        DispatchQueue.main.async {
            self.reportState = .loading
            
            do {
                let payments = try await repository.getPaymentsByDateRange(
                    startDate: Int64(startDate.timeIntervalSince1970),
                    endDate: Int64(endDate.timeIntervalSince1970)
                )
                
                let summary = calculateSummary(payments: payments)
                self.reportState = .success(payments: payments, summary: summary)
            } catch {
                self.reportState = .error("Failed to generate report: \(error.localizedDescription)")
            }
        }
    }
    
    func exportPaymentsToExcel(payments: [Payment], summary: PaymentSummary) {
        DispatchQueue.main.async {
            self.exportState = .loading
            
            do {
                let file = try await excelExporter.exportPaymentsToExcel(
                    payments: payments,
                    summary: summary
                )
                
                if let file = file {
                    self.exportState = .success(file: file)
                } else {
                    self.exportState = .error("Failed to export payment data")
                }
            } catch {
                self.exportState = .error("Export failed: \(error.localizedDescription)")
            }
        }
    }
    
    func exportLoanTracker() {
        DispatchQueue.main.async {
            self.loanTrackerState = .loading
            
            do {
                let file = try await excelExporter.exportLoanTracker()
                
                if let file = file {
                    self.loanTrackerState = .success(file: File)
                } else {
                    self.loanTrackerState = .error("Failed to export loan tracker")
                }
            } catch {
                self.loanTrackerState = .error("Export failed: \(error.localizedDescription)")
            }
        }
    }
    
    func exportWeeklyCollectionReport() {
        DispatchQueue.main.async {
            self.exportState = .loading
            
            do {
                let file = try await excelExporter.exportWeeklyCollectionReport()
                
                if let file = file {
                    self.exportState = .success(file: file)
                } else {
                    self.exportState = .error("Failed to export weekly collection report")
                }
            } catch {
                self.exportState = .error("Export failed: \(error.localizedDescription)")
            }
        }
    }
    
    func resetStates() {
        DispatchQueue.main.async {
            self.exportState = .idle
            self.reportState = .idle
            self.loanTrackerState = .idle
        }
    }
    
    private func calculateSummary(_ payments: [Payment]) -> PaymentSummary {
        let totalAmount = payments.reduce(0) { $0 + $1.amount }
        let totalPayments = payments.count
        let cashPayments = payments.filter { $0.paymentMode == "CASH" }.count
        let phonePayments = payments.filter { $0.paymentMode == "PHONE" }.count
        let duePayments = payments.filter { $0.paymentType == "DUE" }.count
        
        return PaymentSummary(
            totalAmount: totalAmount,
            totalPayments: totalPayments,
            cashPayments: cashPayments,
            phonePayments: phonePayments,
            duePayments: duePayments
        )
    }
}

enum ExportState: Equatable {
    case idle
    case loading
    case success(File)
    case error(String)
}

enum ReportState: Equatable {
    case idle
    case loading
    case success(payments: [Payment], summary: PaymentSummary)
    case error(String)
}

enum LoanTrackerState: Equatable {
    case idle
    case loading
    case success(File)
    case error(String)
}

struct PaymentSummary {
    let totalAmount: Double
    let totalPayments: Int
    let cashPayments: Int
    let phonePayments: Int
    let duePayments: Int
}

File: android/app/src/main/java/com/example/finance/ui/screens/reports/ReportsViewModel.kt
package com.example.finance.ui.screens.reports

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.finance.data.repository.FinanceRepository
import com.example.finance.util.ExcelExporter
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.io.File
import javax.inject.Inject

@HiltViewModel
class ReportsViewModel @Inject constructor(
    @ApplicationContext private val context: Context,
    private val repository: FinanceRepository
) : ViewModel() {

    private val _exportState = MutableStateFlow<ExportState>(ExportState.Idle)
    val exportState = _exportState.asStateFlow()

    private val _reportState = MutableStateFlow<ReportState>(ReportState.Idle)
    val reportState = _reportState.asStateFlow()

    private val _loanTrackerState = MutableStateFlow<LoanTrackerState>(LoanTrackerState.Idle)
    val loanTrackerState = _loanTrackerState.asStateFlow()

    fun generateReport(startDate: Long, endDate: Long) {
        viewModelScope.launch {
            _reportState.value = ReportState.Loading
            try {
                val payments = repository.getPaymentsByDateRange(startDate, endDate)
                val summary = calculateSummary(payments)
                _reportState.value = ReportState.Success(payments, summary)
            } catch (e: Exception) {
                _reportState.value = ReportState.Error("Failed to generate report: ${e.message}")
            }
        }
    }

    fun exportPaymentsToExcel(payments: List<Payment>, summary: PaymentSummary) {
        viewModelScope.launch {
            _exportState.value = ExportState.Loading
            try {
                val exporter = ExcelExporter(context, repository)
                val file = exporter.exportPaymentsToExcel(payments, summary)
                if (file != null) {
                    _exportState.value = ExportState.Success(file)
                } else {
                    _exportState.value = ExportState.Error("Failed to export payment data")
                }
            } catch (e: Exception) {
                _exportState.value = ExportState.Error("Export failed: ${e.message}")
            }
        }
    }

    fun exportLoanTracker() {
        viewModelScope.launch {
            _loanTrackerState.value = LoanTrackerState.Loading
            try {
                val exporter = ExcelExporter(context, repository)
                val file = exporter.exportLoanTracker()
                if (file != null) {
                    _loanTrackerState.value = LoanTrackerState.Success(file)
                } else {
                    _loanTrackerState.value = LoanTrackerState.Error("Failed to export loan tracker")
                }
            } catch (e: Exception) {
                _loanTrackerState.value = LoanTrackerState.Error("Export failed: ${e.message}")
            }
        }
    }

    private fun calculateSummary(payments: List<Payment>): PaymentSummary {
        val totalAmount = payments.sumOf { it.amount }
        val totalPayments = payments.size
        val cashPayments = payments.count { it.paymentMode == "CASH" }
        val phonePayments = payments.count { it.paymentMode == "PHONE" }
        val duePayments = payments.count { it.paymentType == "DUE" }
        
        return PaymentSummary(
            totalAmount = totalAmount,
            totalPayments = totalPayments,
            cashPayments = cashPayments,
            phonePayments = phonePayments,
            duePayments = duePayments
        )
    }

    fun resetStates() {
        _exportState.value = ExportState.Idle
        _reportState.value = ReportState.Idle
        _loanTrackerState.value = LoanTrackerState.Idle
    }
}

sealed class ExportState {
    object Idle : ExportState()
    object Loading : ExportState()
    data class Success(val file: File) : ExportState()
    data class Error(val message: String) : ExportState()
}

sealed class ReportState {
    object Idle : ReportState()
    object Loading : ReportState()
    data class Success(val payments: List<Payment>, val summary: PaymentSummary) : ReportState()
    data class Error(val message: String) : ReportState()
}

sealed class LoanTrackerState {
    object Idle : LoanTrackerState()
    object Loading : LoanTrackerState()
    data class Success(val file: File) : ExportState()
    data class Error(val message: String) : LoanTrackerState()
}

data class PaymentSummary(
    val totalAmount: Double,
    val totalPayments: Int,
    val cashPayments: Int,
    val phonePayments: Int,
    val duePayments: Int
)
